packages/public-api/src/website/service.ts

```diff
@@ -1,80 +1,120 @@
-    getRepository(WebPage).findOne({ url }),
-    internalApi.initiateUrlProcessing(url),
-    // Clear cached results
-    getRepository(WebPagePackage).delete({ hostname: getHostnameFromUrl(url) }),
-  ]);
-
-  // Insert or update database entry
-  return getRepository(WebPage).save({
-    url: internal.url,
-    hostname: getHostnameFromUrl(internal.url),
-    status: mapInternalWebsiteStatus(internal.status),
-    ...(cached ? { id: cached.id } : {}),
-  });
-}
+import {
+  WebPageScan,
+  getDatabaseConnection,
+  internalApi,
+  Hostname,
+  WebPage,
+} from '@gradejs-public/shared';
+import { EntityManager } from 'typeorm';
+import { syncPackageUsageByHostname } from '../projections/syncPackageUsageByHostname';
+import { syncScansWithVulnerabilities } from '../projections/syncScansWithVulnerabilities';
+import { SystemApi } from '../systemApiRouter';
 
-export async function syncWebPage(webpage: WebPage) {
-  // Skip sync for processed items
-  if (webpage.status !== WebPage.Status.Pending) {
-    return;
-  }
+const RESCAN_TIMEOUT_MS = 1000 * 60 * 60 * 24; // 1 day in ms
 
-  const internal = await internalApi.fetchUrlPackages(webpage.url);
-  const nextStatus = mapInternalWebsiteStatus(internal.status);
+export async function findOrCreateWebPage(url: URL, em: EntityManager) {
+  const hostnameRepo = em.getRepository(Hostname);
+  const webPageRepo = em.getRepository(WebPage);
 
-  // Save updated status if changed
-  if (webpage.status !== nextStatus) {
-    webpage.status = nextStatus;
-    await webpage.save();
+  let hostnameEntity = await hostnameRepo.findOne({
+    hostname: url.hostname,
+  });
+  if (!hostnameEntity) {
+    hostnameEntity = await hostnameRepo.save({
+      hostname: url.hostname,
+    });
   }
 
-  // Save packages if ready
-  if (webpage.status === WebPage.Status.Processed && internal.detectedPackages.length > 0) {
-    await getRepository(WebPagePackage).upsert(
-      internal.detectedPackages.map((pkg) => ({
-        latestUrl: internal.url,
-        hostname: getHostnameFromUrl(internal.url),
-        packageName: pkg.name,
-        possiblePackageVersions: pkg.possibleVersions,
-        packageVersionRange: pkg.versionRange,
-        packageMetadata: {
-          approximateByteSize: pkg.approximateSize ?? undefined,
-        },
-      })),
-      ['hostname', 'packageName']
-    );
+  let webPageEntity = await webPageRepo
+    .createQueryBuilder('web_page')
+    .where('web_page.hostname_id = :hostnameId', { hostnameId: hostnameEntity.id })
+    .andWhere('web_page.path = :path', { path: url.pathname })
+    .limit(1)
+    .getOne();
+  if (!webPageEntity) {
+    webPageEntity = await webPageRepo.save({
+      hostname: hostnameEntity,
+      path: url.pathname,
+    });
   }
+
+  return webPageEntity;
 }
 
-export async function getWebPagesByHostname(hostname: string) {
-  return getRepository(WebPage).find({ hostname });
+export async function requestWebPageScan(url: string) {
```

Comment:
According to the naming I wouldn't expect that function returns a cached scan. Can we rename it to something like `getOrRequestWebPageScan`?